prusti-specs/src/specifications/common.rs

```diff
@@ -187,6 +187,13 @@ pub struct ForAllVars<EID, AT> {
     pub vars: Vec<AT>,
 }
 
+#[derive(Debug, Clone)]
+pub struct SpecEntVars<EID, AT> {
+    pub spec_id: SpecificationId,
+    pub id: EID,
+    pub vars: Vec<AT>,
+}
```

Comment:
I wonder if we really need a separate `struct` for this? `TypeCond` also uses `ForAllVars`. Maybe we could just rename the struct to something like `SpecVars` and use it for all three?

Reply:
I agree that this version of `SpecEntVars` looked superfluous. But keep in mind that type checking for closure specifications is slightly more intricate than for quantifiers, because we can't type-check the pre- and postconditions in one go: The postcondition may refer to `result` but the precondition may not. For this reason, my current implementation defines `SpecEntVars` like this:
```rust
#[derive(Debug, Clone)]
pub struct SpecEntVars<EID, AT> {
    pub spec_id: SpecificationId,
    pub pre_id: EID,
    pub post_id: EID,
    pub args: Vec<AT>,
    pub result: AT,
}
```
i.e. we additionally need the result type, and _two_ IDs, one for type-checking the pre- and one for type-checking the postconditions.

To be honest, `result` is currently hard-coded as `result: i32`, but we could change the syntax to `cl |= |i: i32| -> T [ ... ]` to fix this problem. What doesn't work in this setting is higher-order closures, because if `cl` returns a closure type, we can't write that down (as `T`), because it's anonymous. This is unfortunate but seems to be an intrinsic limitation to the way type checking is implemented right now (note that closures can't be generic, so we can't write `|| -> impl Fn () -> i32`, as we could for functions).

Reply:
Ok, if `SpecEntVars` ends up different from `ForAllVars`, it makes sense.

As for the return type, we could indeed make the syntax **optionally** accept `-> X` after the arguments. When parsing, it would still be unambiguous, I believe. The higher-order closure-returning-a-closure situation is interesting. We could pretend that the type annotation is valid even if Rust would not accept it, but I predict this would get rather unwieldy very quickly.

I expect that closures like this are not extremely common. Maybe eventually (as a result of your work or later still) closures will not require explicit type annotation in Prusti, both in the specifications and the definitions. Then we could generate type checks that rely on type inference just like what the user wrote in the first place.